FEATURE: Soft removal garbage collection

nezaniel opened the pull request:
see #5478

## Introduction of the `SoftRemovalGarbageCollector`

Service that detects which soft removals in a content repository can be safely transformed to hard removals
and executes those.

### Invalidation

Triggering of the soft removal garbage collection is done via the `WorkspacePublishingService`

For these commands the contents of the workspace change and afterward there are less pending changes - e.g. less impending conflicts.
Which means that we can trigger the garbage collection in hope to have something cleaned up.

- `PublishWorkspace`
- `PublishIndividualNodesFromWorkspace`
- `DiscardIndividualNodesFromWorkspace`
- `DiscardWorkspace`
- `DeleteWorkspace`
- `RebaseWorkspace`

The command `RebaseWorkspace` is a special case. Because (expect for force-rebase with dropped changes) no changes are expected to be dropped
and thus there are no subtractions from the impending conflicts. Instead, the rebase takes care of synchronizing the new soft removals into the workspace.
New soft removals can either cause new impending conflicts or free the claim of the workspace as the node will no longer be visible.

### Implementation 

The detection/execution is done in the following steps:

1) find all soft removals in the root workspace live `findNodeAggregatesInWorkspaceByExplicitRemovedTag()`

For each soft removal:

2) check if any workspace objects to the transformation
      2.a) by not knowing about the soft removal yet `withVisibleInDependingWorkspacesConflicts()`
      2.b) by having pending changes inside the synchronized soft removal `withImpendingHardRemovalConflicts()`
3) if there are no impending conflicts, execute the hard removal on the root workspace
4) the hard removal will then propagate to the other workspace automatically via rebase

### Sync vs async

In the future we could allow to disable the garbage collection being run synchronously but offer an async job.

### Impediments

In a single user system the garbage collection will immediately turn the users workspace outdated after publishing a non-conflicting removal.

**Upgrade instructions**

none

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-12 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

Commit:
5359 - declare garbage collector and add first test

Commit:
5459 - WIP: Implement basic guardrail tests for garbage collection

mhsdesign committed:
5359 - WIP: Step 0. Garbage collect all soft deletions

mhsdesign committed:
5359 - WIP: Step 1. Garbage collect all soft deletions

if a workspace is outdated and does not have the soft deletion yet, but knows the node we cannot safely remove the node as
the user can still make changes to that node or children any time in the future - or might have done so already.

mhsdesign committed:
5359 - WIP: Step 2. Introduce objections

mhsdesign committed:
5359 - WIP: Step 2.2 Use node aggregates in catchup hook

mhsdesign committed:
5359 - WIP: Step 2.3 Respect soft removal of current node

mhsdesign committed:
5359 - WIP: Step 2.4 Fix node variant handling

fixes Garbage collection will ignore a soft removal if the node has unpublished newly created variants in another workspace

Commit:
5459 - Add test cases for node creation conflicts

Commit:
5459 - properly joint conflict generalizations and add variation conflict tests

mhsdesign committed:
5359 - WIP: Handle node hard deletion conflicts correctly

mhsdesign committed:
5359 - WIP: Build intersection to only check explicit soft removals if interesting

fix: arbage collection will ignore a soft removal if the node has unpublished newly created variants by target in another workspace

mhsdesign committed:
5359 - WIP: Fix `SetNodeReferences` test to expectations

Commit:
5459 - add conflict comparisons to tests

Commit:
5459 - also detect partial conflicts

Commit:
5459 - Adjust test numbers to feature test numbers

Commit:
5459 - Add test cases for successful transformation with potential node variation conflicts

Commit:
5459 - Enhance NodeModificationConflict test cases

Commit:
5459 - streamline names

Commit:
5459 - extend node referencing conflict tests and similarize with modification

Commit:
5459 - extend test cases for subtree tagging

Commit:
5459 - handle nested removals

Commit:
5459 - extend node removal conflict tests

Commit:
5459 - detect impending conflicts on reference targets

Commit:
Test rebase after hard removal

Commit:
note missing test

Commit:
5459 - Extend tests for conflicts between move node and soft removal

mhsdesign committed:
TASK: Use new improved content graph queries

See

https://github.com/neos/neos-development-collection/pull/5482
https://github.com/neos/neos-development-collection/pull/5483

mhsdesign committed:
TASK: Fix flaky tests by sorting the dsps

-        'dimensionSpacePoints' => '[{"example":"source"},{"example":"special"}]'
+        'dimensionSpacePoints' => '[{"example":"special"},{"example":"source"}]'

mhsdesign committed:
TASK: Adjust nested removal test to new query's unordered results

before we used the hierarchy and always run into the problem of deleting the parent first, now with using `findNodeAggregatesTaggedWith` we dont know what to expect.

Only that the order is set by `orderBy('n.relationanchorpoint', 'DESC')` which means that the nodes are ordered by insert - eg. generation of the `relationanchorpoint`

Commit:
5459 - Extend tests for name change conflicting with hard removal

Commit:
5459 - Extend test cases for conflicts between node type change and hard removal

Commit:
5459 - Add failing tests for root node aggregate dimension update conflicts

mhsdesign committed:
TASK: Add test for ImpendingHardRemovalConflictRepository that primary key is not violated

Integrity constraint violation: 1062 Duplicate entry 'default-user-workspace-nodingers-cat' for key 'PRIMARY'

Thats why we use the upsert via `if ($row === false) {`

mhsdesign committed:
TASK: Add test for ImpendingHardRemovalConflictRepository fetched rows are merged correctly

`findAllConflicts` unions the dsps which is tested here

mhsdesign committed:
TASK: Add test that publishing and discarding also flushes the impending conflicts

mhsdesign committed:
TASK: Add notes for catchup hook parent traversal

mhsdesign committed:
TASK: Test that conflicts target all soft removed ancestor nodes

mhsdesign committed:
WIP BUGFIX: Avoid `NodeAggregateDoesCurrentlyNotCoverDimensionSpacePoint` being thrown when deleting

because the child was already removed

mhsdesign committed:
WIP Add debug steps for failing test, `RootNodeAggregateDimensionsWereUpdated` seems to drop the subtree tags on the root node

mhsdesign committed:
TASK: Trigger garbage collection via `WorkspacePublishingService`

Originally the idea was to have a command hook trigger the hard removals on these commands via the proposed `CommandHook::onAfterHandle(): Commands`.

The idea was that on any of those commands the contents of the workspace change and afterward there are less pending changes - e.g. less impending conflicts. Which means that we can trigger the garbage collection in hope to have something cleaned up.

- DeleteWorkspace
- PublishWorkspace
- PublishIndividualNodesFromWorkspace
- DiscardIndividualNodesFromWorkspace
- DiscardWorkspace
- RebaseWorkspace

But a command hook turned out to be not sufficient as we have to catch `handle` in expected cases. Also, by triggering this explicitly we make the code flow more obvious. And we probably will allow to disable this being run synchronously but offer an async job - for that the `WorkspacePublishingService` is a good place to coordinate that too.

The command `RebaseWorkspace` is a special case. Because (expect force-rebases with dropped changes) no changes are expected to be dropped and thus there are no subtractions from the impending conflicts. Instead, the rebase takes care of synchronizing the new soft removals into the workspace which then might introduce new impending conflicts. But a rebase also changes condition 1 as in the case of no changes the soft deletion will be rebased which is a necessity for condition 1 as otherwise no hard removal will be executed when the node is still visible in a workspace.

-----------

In a single user system this will mean that after hitting publish of a deletion the hard deletion is run live and the user workspace will be immediately outdated. This is acceptable and the neos ui is well-prepared to synchronise which is not even necessary for this case.

-----------

A future optimisation, if we prefer staying synchronous and optimising that would be to check that the invalidation is only triggered if the impending conflicts for the $command->workspace did change, as otherwise we don't expect any allowed soft deletion removals.
Even more fast could be to just check the node ids where we have less impending conflicts after the publish operation and just trigger flushing just for those.
Though overall moving this in an async que is for huge sites probably the best optimisation to remove load from editors requests.

mhsdesign committed:
BUGFIX: Hard remove nodes where specialisation has been removed before

Previously we subtracted the objection directly which doesnt allow to later figure out if a dimensions space point conflicts or has been removed already. Now to answer that we have to track both information: Which dimensions a node was soft removed in and what the objections are.

Solves: invert logic: compare specialization set with impending conflicts

mhsdesign committed:
TASK: Garbage collect specializations when generalisation is still in conflict

instead of waiting until the generalisation is out of conflict and  removing all nodes at once we instead cleanup the specialisations where possible already

mhsdesign committed:
TASK: Avoid obscure error when attempting to remove node aggregate via `allSpecializations`

> DimensionSpacePointNotFound: [] was not found in the allowed dimension subspace

Is thrown when `getSpecializationSet()` is invoked because root node only cover [] which is always returned by: getOccupationByCovered()

When the command RemoveNodeAggregate is executed with payload:
  | Key                          | Value                    |
  | nodeAggregateId              | "lady-eleonode-rootford" |
  | coveredDimensionSpacePoint   | {"example":"source"}     |
  | nodeVariantSelectionStrategy | "allSpecializations"     |

mhsdesign committed:
TASK: Avoid garbage collecting root node aggregates because they cant be removed easily

Avoid obscure error when attempting to remove node aggregate via `allSpecializations`

> DimensionSpacePointNotFound: [] was not found in the allowed dimension subspace

# its not possible to remove a root node without using allVariants
When the command RemoveNodeAggregate is executed with payload:
  | Key                          | Value                    |
  | nodeAggregateId              | "lady-eleonode-rootford" |
  | coveredDimensionSpacePoint   | {"example":"source"}     |
  | nodeVariantSelectionStrategy | "allSpecializations"     |

mhsdesign committed:
TASK: Avoid traversing all parents if the node aggregate is not soft removed

Possible via the new `getCoveredDimensionsTaggedBy` https://github.com/neos/neos-development-collection/pull/5484

Also adjustments to the other new cr api's

Commit:
5459 - move dimension space point set reduction to variation graph

Commit:
5459 - move generalization resolution to variation graph

Commit:
5459 - refactor NodeAggregateIdsWithDSPs to explict conflict model

Commit:
5459 - add test for removing partially covering nodes

mhsdesign committed:
TASK: Combine two equal tests where specialisation has already been removed

mhsdesign committed:
TASK: Remove obsolete `getGeneralizationSetForSet`

because by inverting the logic in 6aba628572d288c33b067ceb5c9ff1e41d4f5d3e
we dont have to 'erase' the generalisations anymore but will take care of that in reverse by checking if any attempted generalisation will take any conflicting specialisation with it and then stop the hard removal

mhsdesign committed:
TASK: Adjust naming things

after refactoring from 6aba628572d288c33b067ceb5c9ff1e41d4f5d3e

mhsdesign committed:
TASK: Only check depending on workspace of live for non soft removed nodes

mhsdesign committed:
TASK: Avoid exception thrown in test if live doesnt exist

Caused by the step:

the workspace "some-root-workspace" is deleted

mhsdesign committed:
TASK: Explain workaround because NodeReferencesWereSet $affectedDimensionSpacePoints does not exist

technically we could also consider only conflicting

```php
NodePropertiesWereSet::class => DimensionSpacePointSet::fromArray(array_map(
    // find out where the property change was caused e.g. like NodeReferencesWereSet::$affectedSourceOriginDimensionSpacePoints
    fn (DimensionSpacePoint $coveredDimension) => $nodeAggregate->getOccupationByCovered($coveredDimension)->toDimensionSpacePoint(),
    array_values($eventInstance->affectedDimensionSpacePoints->points)
)),
NodeReferencesWereSet::class => $eventInstance->affectedSourceOriginDimensionSpacePoints->toDimensionSpacePointSet(),
```

And similar also only respect the occupied dimensions in `NodeAggregateTypeWasChanged` but thats a refactoring for later.

But this requires bigger test adjustments right now.

mhsdesign committed:
TASK: Improve documentation and add headers

mhsdesign reviewed (approved):
We paired / worked on this together so my +1 should not justify a merge yet ;)

But i just tested this in combination with https://github.com/neos/neos-development-collection/pull/5463 via the Neos ui and it seems to work as expected.

Make sure that you test it with two users and with just a single user as this results in different behaviour regarding the cleanup.

Also i watched the hierarchy relations and after rebasing all workspaces i could see how the soft removed nodes were removed.

One special case might be the last workspace that is rebased: Now a rebase will lead to the triggering of the garbage collection which - in case you are based on live directly - will turn your workspace out of date again. I considered to put the following re-rebase in `rebaseWorkspace` but this is too magic after all i think. Instead - as its never guaranteed that the workspace is up to date (especially with multiple users) - i think the neos ui should provide a feedback loop here to tell: there were changes just now, rebase again.

I fear loose magic like this could make things harder to debug for us.

```php
$rebasedWorkspace = $contentRepository->findWorkspaceByName($workspaceName);
if ($rebasedWorkspace && $rebasedWorkspace->status === WorkspaceStatus::OUTDATED) {
    // in case the soft removal garbage collection did clean up nodes on the base workspace
    $contentRepository->handle(RebaseWorkspace::create($workspaceName));
}
```

mhsdesign committed:
TASK: Ensure `SoftRemovalGarbageCollector` runs without security restrictions

Commit:
Merge branch '9.0' into 5459-softRemovalGarbageCollection

Commit:
pacify linter

mhsdesign committed:
TASK: Prevent tethered nodes to be soft removed

The new constraint is similar to what https://github.com/neos/neos-development-collection/issues/4821 proposes, as this too disallows a specific tag ("disabled") to be placed on tethered nodes.

We discussed to prevent general tagging of tethered nodes but this will make it hard to build for example a member tagged area where the member tethered child node from the site should be tagged so.

mhsdesign committed:
TASK: Add missing `internal` declarations

mhsdesign committed:
TASK: Also constraint soft removing root node aggregates

Root nodes can be tagged in a variety of specialisations via `STRATEGY_ALL_SPECIALIZATIONS` but can
only be removed in all variants. This conflicts with the dimension handling of the garbage collector.

Further soft removing the sites root node is not a Neos use case, and not possible out of the box

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 5459-softRemovalGarbageCollection

mhsdesign committed:
TASK: Use new `NeosSubtreeTag::removed()` instead of temporary stub

as per https://github.com/neos/neos-development-collection/pull/5463

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 5459-softRemovalGarbageCollection

mhsdesign committed:
TASK: Adjust to use `EmbedsWorkspaceName`

mhsdesign committed:
TASK: Move introduced `SoftRemoval` into new `SubtreeTagging` namespace

mhsdesign committed:
TASK: Combine `SoftRemovalConstraints` into `NeosSubtreeTaggingConstraintChecks`

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 5459-softRemovalGarbageCollection

mhsdesign committed:
Jahaha: `match` always needs a space ... ill never get that right
